### PR TITLE
Initial version of MongoDB instrumentation

### PR DESCRIPTION
**Pull request by lstrojny**
Uses the MongoDB APM subscriber infrastructure instead of hooks, so it doesn’t even require the `opentelemetry` extension.

The sub-project setup was blatantly stolen from the PDO module so I am 100% I am missing something.

### CODE REVIEW

File: src/Instrumentation/MongoDB/_register.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+declare(strict_types=1);
+
+use OpenTelemetry\Contrib\Instrumentation\MongoDB\MongoDBInstrumentation;
+use OpenTelemetry\SDK\Sdk;
+
+if (class_exists(Sdk::class) && Sdk::isInstrumentationDisabled(MongoDBInstrumentation::NAME) === true) {
+    return;
+}
+
+if (extension_loaded('opentelemetry') === false) {
```
Comment (lstrojny) [inline src/Instrumentation/MongoDB/_register.php:12]: Note to self: this can go

File: src/Instrumentation/MongoDB/phpstan.neon.dist [hunk 2]
```diff
@@ -0,0 +1,9 @@
+includes:
+    - vendor/phpstan/phpstan-phpunit/extension.neon
+
+parameters:
+    tmpDir: var/cache/phpstan
+    level: 5
```
Comment (lstrojny) [inline src/Instrumentation/MongoDB/phpstan.neon.dist:6]: Increase level

File: src/Instrumentation/MongoDB/psalm.xml.dist [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<psalm
+    errorLevel="3"
```
Comment (lstrojny) [inline src/Instrumentation/MongoDB/psalm.xml.dist:3]: Make stricter

**Comment by brettmc** [comment]
Welcome, @lstrojny 
I've taken a first pass at the code, it looks good to me.
I notice the tests are failing for php < 8.2, will be run in lower versions? If not, can you add exclusions to the workflow to disable tests against earlier versions? If possible, it would be good if it could work with as many versions of PHP as we support (7.4+, although most auto-instrumentations only support 8+)

**Comment by lstrojny** [comment]
@brettmc fixed. Supporting 7.4 and up is trivially possible so let’s do that for now. Enums in PHP 8.1 is probably the only interesting feature, especially for the OTEL core libraries. Adopting those will probably require another major version so we can kick that can down the road.

**Comment by brettmc** [comment]
@lstrojny making progress. Integration tests are failing due to no mongodb available.
We can either look to disable those tests, or add a mongo (which is probably a better test).

I had a quick play and it looks like you can conditionally install mongo, so if you try adding a step into `.github/workflows/php.yml` like so, it may just work:

```
    - name: (optional) Setup mongodb
      uses: supercharge/mongodb-github-action@1.8.0
      if: matrix.project == 'Instrumentation/MongoDB'
      with:
        mongodb-version: 6.0
```

**Comment by lstrojny** [comment]
We shouldn't even need a real MongoDB instance as the attempt to query should be good enough. Can you retry?

**Comment by brettmc** [comment]
@lstrojny I did a little more investigation into mongodb. It should be available on localhost:27017, however in my testing there is a race condition: it looks like the tests are running before mongo is ready to accept connections.
I raised an issue against the action, but in the meantime I found that adding a replicaset does have a readiness check, so try adding this to the `with` section: `mongodb-replica-set: test-rs`

File: src/Instrumentation/MongoDB/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
+# OpenTelemetry MongoDB auto-instrumentation
+
+**Preferred and simplest way to install auto-instrumentation (c extension plus instrumentation libraries) is to use [opentelemetry-instrumentation-installer](https://github.com/open-telemetry/opentelemetry-php-contrib/tree/main/src/AutoInstrumentationInstaller).**
```
Comment (xvilo) [inline src/Instrumentation/MongoDB/README.md:3]: Do we need the C-Extension or could we leave this out?

Comment (lstrojny) [reply src/Instrumentation/MongoDB/README.md:3]: We don’t need it, no. So we can leave that part out

Comment (lstrojny) [reply src/Instrumentation/MongoDB/README.md:3]: Thought about it more: we could leave it out, but then people wouldn’t be pointed to the auto installer and wouldn't install the extension. So instead I added a comment about standalone integration but still mentioned the installer as the preferred way.

**Review by brettmc — APPROVED** [review]

**Comment by brettmc** [comment]
Released to packagist as v0.0.1

**Pull request merged**